### PR TITLE
linux: overlayfs support

### PR DESCRIPTION
**Pull request by freqlabs**
Backported https://github.com/openzfs/zfs/pull/9600 and https://github.com/openzfs/zfs/pull/12209 with new feature flags removed.

This means in cases where a pool is not cleanly exported you may end up
with loss of unflushed TX_RENAME_EXCHANGE and TX_RENAME_WHITEOUT
transactions and all transactions that follow them in the ZIL for that
dataset if the pool is then imported by a release build of ZFS that
does not recognize the new TX types.  A warning message is logged to
dmesg and the pool is fully usable, just missing the last changes.
Debug builds will panic when they fail to replay the full log.

The alternative is a pool that cannot be imported on any version of ZFS
lacking support for TX_RENAME_EXCHANGE and TX_RENAME_WHITEOUT replay
when uncleanly exported.  That prevents people from either upgrading
their pool for the ability to use overlayfs features or going back to a
previous boot environment after having upgraded the pool.

Jira: https://ixsystems.atlassian.net/browse/NAS-109036

### CODE REVIEW

File: module/os/linux/zfs/zfs_vnops_os.c [hunk 1]
```diff
@@ -2986,58 +3056,110 @@ zfs_rename(znode_t *sdzp, char *snm, znode_t *tdzp, char *tnm,
...
+		zfs_log_rename_exchange(zilog, tx,
+		    (flags & FIGNORECASE ? TX_CI : 0), sdzp, sdl->dl_name,
+		    tdzp, tdl->dl_name, szp);
+		break;
+	case RENAME_WHITEOUT:
+		zfs_log_rename_whiteout(zilog, tx,
+		    (flags & FIGNORECASE ? TX_CI : 0), sdzp, sdl->dl_name,
+		    tdzp, tdl->dl_name, szp, wzp);
+		break;
+	default:
```
Comment (ixhamza) [inline module/os/linux/zfs/zfs_vnops_os.c:3151]: ```rflags = 0``` for ```rename()``` system call will also fallback to default case. :thinking:

Comment (freqlabs) [reply module/os/linux/zfs/zfs_vnops_os.c:3151]: I think that's the idea, `rename()` does `zfs_log_rename()`.

File: module/os/freebsd/zfs/zfs_vnops_os.c [hunk 2]
```diff
@@ -3473,14 +3473,17 @@ zfs_do_rename_impl(vnode_t *sdvp, vnode_t **svpp, struct componentname *scnp,
...
-    cred_t *cr, int flags)
+    cred_t *cr, int flags, uint64_t rflags, vattr_t *wo_vap)
 {
 	struct componentname scn, tcn;
 	vnode_t *sdvp, *tdvp;
 	vnode_t *svp, *tvp;
 	int error;
 	svp = tvp = NULL;
 
+	if (rflags != 0 || wo_vap != NULL)
```
[the remark this answers is not in the document]
Comment (freqlabs) [reply module/os/freebsd/zfs/zfs_vnops_os.c:3484]: `rflags` should always be zero and `wo_vap` always null on FreeBSD. We don't handle exchange or whiteout.

File: AUTHORS [hunk 3]
```diff
@@ -19,6 +19,7 @@ CONTRIBUTORS:
     Albert Lee <trisk@nexenta.com>
     Alec Salazar <alec.j.salazar@gmail.com>
     Alejandro R. Sedeño <asedeno@mit.edu>
+    Aleksa Sarai <cyphar@cyphar.com>
```
Comment (alek-p) [inline AUTHORS:22]: this contributor should be on the next line

Comment (freqlabs) [reply AUTHORS:22]: I'll relay this to the upstream PR

File: module/zfs/zfs_replay.c [hunk 4]
```diff
@@ -663,13 +666,104 @@ zfs_replay_rename(void *arg1, void *arg2, boolean_t byteswap)
...
+		byteswap_uint64_array(lr, sizeof (*lr));
+
+	return (do_zfs_replay_rename(zfsvfs, lr, sname, tname, RENAME_EXCHANGE,
+	    NULL));
+#else
+	/*
+	 * We should never reach this point because the feature is not
+	 * supported by non-Linux versions of OpenZFS.
+	 */
+	PANIC("TX_RENAME_EXCHANGE cannot be replayed on non-Linux systems.");
```
Comment (freqlabs) [inline module/zfs/zfs_replay.c:710]: Should remove this panic.

File: module/zfs/zfs_replay.c [hunk 5]
```diff
@@ -663,13 +666,104 @@ zfs_replay_rename(void *arg1, void *arg2, boolean_t byteswap)
...
+		return (error);
+
+	return (do_zfs_replay_rename(zfsvfs, &lr->lr_rename, sname, tname,
+	    RENAME_WHITEOUT, &xva.xva_vattr));
+#else
+	/*
+	 * We should never reach this point because the feature is not
+	 * supported by non-Linux versions of OpenZFS.
+	 */
+	PANIC("TX_RENAME_WHITEOUT cannot be replayed on non-Linux systems.");
```
Comment (freqlabs) [inline module/zfs/zfs_replay.c:762]: Should remove this panic.

**Comment by freqlabs** [comment]
I reworked the backport of the commit adding the new TX types so they line up with the numbers in master (because we don't have TX_SETSAXATTR backported), and I adjusted the replay vectors to include an error function in the missing locations.

**Comment by freqlabs** [comment]
Fixed zdb to handle the hole for setsaxattr, and dropped manpage changes for the features.

**Review by amotin — APPROVED** [review]

**Pull request merged**